infinite-tracing/src/main/java/com/newrelic/ConnectionSynchronizer.java

```diff
@@ -0,0 +1,116 @@
+package com.newrelic;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Provides utilities for communicating connection status information across the various threads responsible for
+ * connecting and responding to connection errors.
+ *
+ * This class is thread safe.
+ */
+class ConnectionSynchronizer {
```

Comment:
This class is a rewrite of `ConnectionStatus` and represents the bulk of the change of this PR. 

The contract of the infinite trace gRPC server is a bidirection stream, with the client sending spans, and the server sending response status messages. Sending spans happens in a single thread initialized in `SpanEventConsumer`. This thread also ends up managing the connection to the gRPC server. The responses sent by the server are handled by threads of cached thread pool, which is the default behavior of the `ManagedChannel` produced by `ChannelFactory#createChannel`.

This class's role is to synchronize connection status between these threads. There are two main improvements in this implementation of its `ConnectionStatus` predecessor:
1. The `awaitRequiresReconnect` method, equivalent to `ConnectionStatus#blockOnConnection`, uses a latch to block instead of busy waiting.
2. The API is simplified, with the notable removal of enum responses for callers to handle.